apps/module-presale/contracts/Presale.sol

```diff
@@ -0,0 +1,336 @@
+pragma solidity ^0.4.24;
+
+import "@aragon/os/contracts/apps/AragonApp.sol";
+
+import "@aragon/os/contracts/lib/math/SafeMath.sol";
+import "@aragon/os/contracts/lib/math/SafeMath64.sol";
+
+import "@aragon/os/contracts/lib/token/ERC20.sol";
+import "@aragon/apps-shared-minime/contracts/MiniMeToken.sol";
+import "@aragon/apps-token-manager/contracts/TokenManager.sol";
+
+
+contract Presale is AragonApp {
+    using SafeMath for uint256;
+    using SafeMath64 for uint64;
+
+    /*
+     * Events
+     */
+
+    event SaleStarted();
+    event SaleClosed();
+    event TokensPurchased(address indexed buyer, uint256 daiSpent, uint256 tokensPurchased, uint256 purchaseId);
+    event TokensRefunded(address indexed buyer, uint256 daiRefunded, uint256 tokensBurned, uint256 purchaseId);
+
+    /*
+     * Errors
+     */
+
+    string private constant ERROR_INVALID_STATE                  = "PRESALE_INVALID_STATE";
+    string private constant ERROR_INSUFFICIENT_DAI_ALLOWANCE     = "PRESALE_INSUFFICIENT_DAI_ALLOWANCE";
+    string private constant ERROR_INSUFFICIENT_DAI               = "PRESALE_INSUFFICIENT_DAI";
+    string private constant ERROR_INVALID_TOKEN_CONTROLLER       = "PRESALE_INVALID_TOKEN_CONTROLLER";
+    string private constant ERROR_NOTHING_TO_REFUND              = "PRESALE_NOTHING_TO_REFUND";
+    string private constant ERROR_DAI_TRANSFER_REVERTED          = "PRESALE_DAI_TRANSFER_REVERTED";
+    string private constant ERROR_INVALID_DAI_TOKEN              = "PRESALE_INVALID_DAI_TOKEN";
+    string private constant ERROR_INVALID_TIME_PERIOD            = "PRESALE_INVALID_TIME_PERIOD";
+    string private constant ERROR_INVALID_DAI_FUNDING_GOAL       = "PRESALE_INVALID_DAI_FUNDING_GOAL";
+    string private constant ERROR_INVALID_PERCENT_VALUE          = "PRESALE_INVALID_PERCENT_VALUE";
+    string private constant ERROR_INVALID_POOL                   = "PRESALE_INVALID_POOL";
+    string private constant ERROR_INVALID_BENEFICIARY_ADDRESS    = "PRESALE_INVALID_BENEFICIARY_ADDRESS";
+    string private constant ERROR_EXCEEDS_FUNDING_GOAL           = "PRESALE_EXCEEDS_FUNDING_GOAL";
+
+    /*
+     * Roles
+     */
+
+    bytes32 public constant START_ROLE = keccak256("START_ROLE");
+    bytes32 public constant BUY_ROLE   = keccak256("BUY_ROLE");
```

Comment:
Perhaps there should be a role for ending the presale?

Reply:
My concern with having a role here is that if it is used to restrict who can close the sale, there is no guarantee to participants that this will ever be called. The contract could be stuck in closed state. 

Since there is no optionality in terms of calling the close function, and it can only be called once the pre-sale has entered the closed state, what is the case for having it be a protected function?

Reply:
It would only be useful if we decided to allow the fundraising to be over bought, such that the role holder could close the funding early.

But I agree, let's keep this unprotected.